### ISSUE TITLE
'list' object has no attribute 'queryset'

### ISSUE THREAD

**Issue by stcbus** [yourlabs/django-autocomplete-light#1320]
Using latest dev git install. I have seen this error a few times from previous searches (such as #790 , so maybe it came back?

Simple model called FundraisingTeam with a team name, that I reference as a ForeignKey field here:

```
class EventRegistration(models.Model):
    ...
    fundraising_team = models.ForeignKey(
        FundraisingTeam, on_delete=models.SET_NULL, blank=True, null=True
    )
```

Then, in my view the query (removed the filter to just double check):

```
class FundraisingTeamAutocomplete(autocomplete.Select2QuerySetView):
    def get_queryset(self):
        # Don't forget to filter out results depending on the visitor !

        qs = FundraisingTeam.objects.all()

        if self.q:
            qs = qs.filter(name__istartswith=self.q)

        return qs
```

And finally, I override the get_form() in my createview to set the widget:

```
class RegisterView(generic.edit.CreateView):
    model = EventRegistration
    fields = ["data", "fundraising_team"]
    template_name = "app/temp.html"

    def get_form(self):
        form = super(RegisterView, self).get_form()
        fundraising_enabled = (
            EventInstance.objects.get(pk=self.kwargs.get("event_instance_id"))
        ).enable_fundraising
        if fundraising_enabled:
            form.fields["fundraising_team"].widget = autocomplete.ModelSelect2(
                url=reverse(
                    "endurance:team_autocomplete",
                    kwargs={"pk": self.kwargs.get("event_instance_id")},
                ),
                attrs={
                    # Set some placeholder
                    "data-placeholder": "Type to start a search",
                    # Only trigger autocompletion after 3 characters have been typed
                    "data-minimum-input-length": 3,
                },
            )
        else:
            del form.fields["fundraising_team"]

        return form

```

My example seems pretty vanilla so far from the tutorial, so not sure what's happening. Latest released Django 4.1. 

Appreciate any help!

**Comment by stcbus** [comment]
I'm sure this isn't a priority and/or everyone is very busy, but is it possible to get some help on this? I am trying to use this module for a project with a tight deadline. I appreciate any help you all can provide!

**Comment by Actionb** [comment]
I assume you're running into this exception? (please, always include a stack trace if you have one)
```python
[...] in filter_choices_to_render
    self.choices.queryset = self.choices.queryset.filter(
AttributeError: 'list' object has no attribute 'queryset'
```

The formfield sets the choices on the widget when the [formfield is instantiated](https://github.com/django/django/blob/03bc92af978f2148c1141673c730f17772cd7037/django/forms/models.py#L1492):

```python
    def _set_queryset(self, queryset):
        self._queryset = None if queryset is None else queryset.all()
        self.widget.choices = self.choices
```

But you override the widget after and never set the choices - so this widget has the default value for `choices`, [an empty list](https://github.com/django/django/blob/03bc92af978f2148c1141673c730f17772cd7037/django/forms/fields.py#L877). 

So either explicitly set the choices on the widget after it has been created in your `get_form` method, or declare the widget via `form.Meta.widgets` [like described](https://django-autocomplete-light.readthedocs.io/en/master/tutorial.html#use-the-view-in-a-form-widget) and let Django take care of that. I suggest the latter.

**Comment by stcbus** [comment]
Thanks, so much @Actionb - not sure how I missed that difference. I blame being a newbie and also seeing another issue with it in the view that I lost context when copying/pasting.

Checking now, I am actually forwarding 2 values so that I can create objects, including a kwarg. How would I do this in the model meta without access to self?

```
 form.fields["fundraising_team"].widget = autocomplete.ListSelect2(
                url=reverse(
                    "endurance:team_autocomplete", kwargs={"pk": event_instance.id}
                ),
                forward=["email", forward.Const(event_instance.id, "event_id")],
                attrs={
                    # Set some placeholder
                    "data-placeholder": "Type to start a search",
                    # Only trigger autocompletion after 3 characters have been typed
                },
            )
```

One final thing - I am actually wondering what do i gain/lose by using the ModelSelect vs ListSelect if I have it working now?

**Comment by Actionb** [comment]
I don't quite follow what you are trying to do with the limited information you have given  - and I don't think the issue tracker is the right place to ask for implementation advice like that (try asking on stackoverflow). 
If you want to keep your current code, you can set the widget's choices after declaring it in `get_form` (https://github.com/yourlabs/django-autocomplete-light/issues/790#issuecomment-276309980) and it should work.

> One final thing - I am actually wondering what do i gain/lose by using the ModelSelect vs ListSelect if I have it working now?

I **think** (I'm not a dal developer) the choice of widget only matters for the initial rendering of the form. After that, the choices are served by your autocomplete view. That's why the choices doesn't _seem_ to matter once the form is up and rendered.

It comes down to what type of object `choices` is.
The default formfield for many-to-one and many-to-many model fields is an instance of `ModelChoiceField`.
To enable lazy evaluation of the choices queryset (and to facilitate enumerating over the queryset items), the formfield ['wraps' the widget choices](https://github.com/django/django/blob/03bc92af978f2148c1141673c730f17772cd7037/django/forms/models.py#L1512) in a `ModelChoiceIterator` object.  

In the dal workflow, `ModelSelect2` [can handle that kind](https://github.com/yourlabs/django-autocomplete-light/blob/62ecb5f2200d31c776ccac7bbd06c31c12be58f2/src/dal/widgets.py#L187) of object whereas `ListSelect` [cannot](https://github.com/yourlabs/django-autocomplete-light/blob/62ecb5f2200d31c776ccac7bbd06c31c12be58f2/src/dal/widgets.py#L71). 
So if you want to override the widget of a `ModelChoiceField` in `Form.Meta.widgets`, you should use a `ModelSelect2` widget, because a `ListSelect` widget won't be able to handle the choices set by the formfield.

**Comment by yangwk4464** [comment]
I encounter the same issue, is that because of some logic conflicts with django-filter?

**Comment by Actionb** [comment]
What do you mean by conflict? This issue comes from overriding the widget of a ModelChoiceField formfield after the field has already been instantiated. The ModelChoiceField sets the widget's choices only once. **If you override the widget after that, you have to set the choices yourself.** Or use the form's meta options to override the widget for a given formfield, as described by the [Django docs](https://docs.djangoproject.com/en/4.2/topics/forms/modelforms/#overriding-the-default-fields).
